New Resource `azurerm_maintenance_configuration`

njuCZ opened the pull request:
partially address #6037 
Once it is merged, I'll go on implementing `azurerm_maintenance_assignment`

After tested, I found the `resource_group_name` from the response of read api is always small case, so I use `azure.SchemaResourceGroupNameDiffSuppress()`

Another point is the key of `tags` from the response of read api is always small case


![image](https://user-images.githubusercontent.com/2786738/76274976-12161700-62bd-11ea-919d-3ed106a60560.png)

katbyte commented on azurerm/internal/services/maintenance/data_source_maintenance_configuration.go:
```diff
@@ -0,0 +1,78 @@
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Could we add some validation here?

njuCZ replied:
I added basic validation for the name: `validation.StringIsNotEmpty`
Because there seems no rule, I tried single `1` as the name, it can succeeded too

katbyte commented on azurerm/internal/services/maintenance/resource_arm_maintenance_configuration.go:
```diff
@@ -0,0 +1,168 @@
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
```
Could we open an issue on the SDK about the casing being returned wrong and link to it here?

katbyte commented on azurerm/internal/services/maintenance/resource_arm_maintenance_configuration.go:
```diff
@@ -0,0 +1,168 @@
+				Optional: true,
+				Default:  "All",
+				ValidateFunc: validation.StringInSlice([]string{
+					"All",
```
Could we use the `maintenance.Scope` constants here instead of strings?

katbyte commented on azurerm/internal/services/maintenance/resource_arm_maintenance_configuration.go:
```diff
@@ -0,0 +1,168 @@
+			MaintenanceScope: maintenance.Scope(scope),
+			Namespace:        utils.String("Microsoft.Maintenance"),
+		},
+		Tags: tags.Expand(t),
```
We can just get the values inline
```suggestion
		Tags: tags.Expand(d.Get("tags").(map[string]interface{})),
```

katbyte commented on azurerm/internal/services/maintenance/resource_arm_maintenance_configuration.go:
```diff
@@ -0,0 +1,168 @@
+		return fmt.Errorf("Error retrieving MaintenanceConfiguration %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	if resp.ID == nil {
```
could we also check for an empty string?

katbyte commented on azurerm/internal/services/maintenance/tests/resource_arm_maintenance_configuration_test.go:
```diff
@@ -0,0 +1,232 @@
+					resource.TestCheckResourceAttr(data.ResourceName, "tags.%", "0"),
+				),
+			},
+			{
```
could we add import steps after every step?

katbyte commented on azurerm/internal/services/maintenance/tests/resource_arm_maintenance_configuration_test.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestRG-%d"
```
Could we use
```suggestion
  name     = "acctestRG-maint-%d"
```

katbyte commented on azurerm/internal/services/maintenance/tests/resource_arm_maintenance_configuration_test.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+resource "azurerm_maintenance_configuration" "test" {
+  name                = "acctest-mc%d"
```
could we ensure uppercase works
```suggestion
  name                = "acctest-MC%d"
```

katbyte commented on azurerm/internal/services/maintenance/tests/resource_arm_maintenance_configuration_test.go:
```diff
@@ -0,0 +1,232 @@
+  scope               = "Host"
+
+  tags = {
+    env = "test"
```
could we ensure uppercase works?
```suggestion
    EnV = "TesT"
```

njuCZ replied:
I have changed the value to `TesT`
as for the key, this service will return small case key, regardless of the input case.
If we set `EnV = "TesT"`, the service will return `env = "TesT"`

njuCZ commented:
@katbyte Thanks for the suggestion, I have refined the codes accordingly. Please have a look again

njuCZ commented:
@jackofallops I have added some support for case insensitive resource ID, and added some notes in the doc, do you think it's ok? The service team says it's by design to make the case insensitive

katbyte commented on azurerm/internal/services/maintenance/resource_arm_maintenance_configuration.go:
```diff
@@ -0,0 +1,169 @@
+			},
+
+			// can not contain upper case key
+			"tags": tags.Schema(),
```
agreed

njuCZ commented:
@katbyte I have added custom validation for the tags and opened an issue, Could you please have a look again?

katbyte commented on azurerm/internal/services/maintenance/parse/maintenance_configuration.go:
```diff
@@ -0,0 +1,39 @@
+		ResourceGroup: id.ResourceGroup,
+	}
+
+	if name, err := id.PopSegment("maintenanceconfigurations"); err != nil {
+		if name, err = id.PopSegment("maintenanceConfigurations"); err != nil {
```
We should open an SKD issue here and link to it

njuCZ replied:
thank you. I have added such bug about returned resource id:
https://github.com/Azure/azure-rest-api-specs/issues/8653
I have noted in the `resource_group_name` schema defnition
